pkg/virtualKubelet/nodeProvider/reconciler.go

```diff
@@ -0,0 +1,423 @@
+package nodeprovider
+
+import (
+	"context"
+	"errors"
+
+	v1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/util/slice"
+
+	netv1alpha1 "github.com/liqotech/liqo/apis/net/v1alpha1"
+	sharingv1alpha1 "github.com/liqotech/liqo/apis/sharing/v1alpha1"
+	advertisementOperator "github.com/liqotech/liqo/internal/advertisementoperator"
+)
+
+// The reconciliation function; every time this function is called,
+// the node status is updated by means of r.updateFromResourceOffer.
+func (p *LiqoNodeProvider) reconcileNodeFromResourceOffer(event watch.Event) error {
+	var resourceOffer sharingv1alpha1.ResourceOffer
+	unstruct, ok := event.Object.(*unstructured.Unstructured)
+	if !ok {
+		return errors.New("error in casting ResourceOffer")
+	}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstruct.Object, &resourceOffer); err != nil {
+		klog.Error(err)
+		return err
+	}
+
+	if event.Type == watch.Deleted || !resourceOffer.DeletionTimestamp.IsZero() {
+		p.updateMutex.Lock()
+		defer p.updateMutex.Unlock()
+		klog.Infof("resourceOffer %v is going to be deleted... set node status not ready", resourceOffer.Name)
+		no, err := p.client.CoreV1().Nodes().Get(context.TODO(), p.nodeName, metav1.GetOptions{})
+		if err != nil {
+			klog.Error(err)
+			return err
+		}
+		for i, condition := range no.Status.Conditions {
+			if condition.Type == v1.NodeReady {
+				no.Status.Conditions[i].Status = v1.ConditionFalse
+				p.onNodeChangeCallback(no)
+				break
+			}
+		}
+
+		if err := p.handleResourceOfferDelete(&resourceOffer); err != nil {
+			klog.Errorf("something went wrong during resourceOffer deletion - %v", err)
+			return err
+		}
+		return nil
+	}
+
+	for {
+		if err := p.updateFromResourceOffer(&resourceOffer); err == nil {
+			klog.Info("node correctly updated from resourceOffer")
+			break
+		} else {
+			klog.Errorf("node update from resourceOffer %v failed for reason %v; retry...", resourceOffer.Name, err)
+		}
+	}
```

Comment:
This risks to be a busy loop in case the error is never solved.
1) I would add an exponential back-off sleep mechanism;
2) I would check whether it is ok if this function never returns trying to solve the error, hence blocking this thread.
The same applies to the same logic below

Reply:
I moved the retry logic with an exponential backoff in the informer handler funcs, can you please see if does it make sense in your opinion?

Reply:
Looks fine to me